File: mobile/flutter/lib/services/deep_link_service.dart
/// deep_link_service.dart
/// Handles Universal Links (iOS) and App Links (Android) for RemitFlow Flutter app.
/// Uses app_links package (add to pubspec.yaml: app_links: ^6.0.0).
///
/// Supported deep link patterns:
///   remitflow://send?to=<beneficiaryId>&amount=<amount>&currency=<currency>
///   remitflow://pay?ref=<paymentRef>
///   remitflow://kyc
///   remitflow://wallet
///   remitflow://transaction/<id>
///   https://app.remitflow.com/send?...  (Universal Link / App Link)

import 'dart:async';
import 'package:app_links/app_links.dart';
import 'package:go_router/go_router.dart';
import 'package:flutter/widgets.dart';

class DeepLinkRoute {
  final String path;
  final Map<String, String> queryParams;

  const DeepLinkRoute({required this.path, this.queryParams = const {}});
}

class DeepLinkService {
  static final _appLinks = AppLinks();
  static StreamSubscription<Uri>? _subscription;

  /// Parse a deep link URI into a navigation path.
  static DeepLinkRoute? parseUri(Uri uri) {
    // Normalize universal links
    final host = uri.host;
    final path = uri.path.replaceFirst('/', '');
    final params = uri.queryParameters;

    // Custom scheme: remitflow://send?to=...
    if (uri.scheme == 'remitflow') {
      return _routeFromHostAndParams(host, params);
    }

    // Universal link: https://app.remitflow.com/send?...
    if (host == 'app.remitflow.com' || host == 'remitflow.app') {
      return _routeFromHostAndParams(path, params);
    }

    return null;
  }

  static DeepLinkRoute? _routeFromHostAndParams(
    String segment,
    Map<String, String> params,
  ) {
    switch (segment) {
      case 'send':
        final queryStr = params.entries
            .map((e) => '${e.key}=${Uri.encodeComponent(e.value)}')
            .join('&');
        return DeepLinkRoute(
          path: '/send',
          queryParams: params,
        );
      case 'pay':
        if (params['ref'] == null) return null;
        return DeepLinkRoute(
          path: '/payment-confirm',
          queryParams: params,
        );
      case 'kyc':
        return const DeepLinkRoute(path: '/kyc');
      case 'wallet':
        return const DeepLinkRoute(path: '/wallet');
      default:
        // Check for transaction/<id> pattern
        final txMatch = RegExp(r'^transaction/(.+)$').firstMatch(segment);
        if (txMatch != null) {
          return DeepLinkRoute(
            path: '/transaction/${txMatch.group(1)}',
          );
        }
        return const DeepLinkRoute(path: '/');
    }
  }

  /// Initialize deep link handling.
  /// Call this in your root widget's initState.
  static Future<void> initialize(GoRouter router) async {
    // Handle cold start deep link
    try {
      final initialUri = await _appLinks.getInitialLink();
      if (initialUri != null) {
        final route = parseUri(initialUri);
        if (route != null) {
          WidgetsBinding.instance.addPostFrameCallback((_) {
            _navigateToRoute(router, route);
          });
        }
      }
    } catch (_) {}

    // Handle foreground deep links
    _subscription = _appLinks.uriLinkStream.listen((uri) {
      final route = parseUri(uri);
      if (route != null) _navigateToRoute(router, route);
    });
  }

  static void _navigateToRoute(GoRouter router, DeepLinkRoute route) {
    if (route.queryParams.isEmpty) {
      router.go(route.path);
    } else {
      final queryStr = route.queryParams.entries
          .map((e) => '${e.key}=${Uri.encodeComponent(e.value)}')
          .join('&');
      router.go('${route.path}?$queryStr');
    }
  }

  /// Dispose the deep link subscription.
  static void dispose() {
    _subscription?.cancel();
    _subscription = null;
  }

  /// Generate a shareable send money link.
  static String generateSendMoneyLink({
    required String beneficiaryId,
    double? amount,
    String currency = 'NGN',
  }) {
    final params = <String, String>{'to': beneficiaryId, 'currency': currency};
    if (amount != null) params['amount'] = amount.toStringAsFixed(2);
    final query = params.entries
        .map((e) => '${e.key}=${Uri.encodeComponent(e.value)}')
        .join('&');
    return 'https://app.remitflow.com/send?$query';
  }

  /// Generate a shareable payment request link.
  static String generatePaymentRequestLink(String ref) {
    return 'https://app.remitflow.com/pay?ref=${Uri.encodeComponent(ref)}';
  }
}

File: uis/react-native/src/services/deepLinkService.ts
/**
 * deepLinkService.ts
 * Handles Universal Links (iOS) and App Links (Android) for RemitFlow.
 *
 * Supported deep link patterns:
 *   remitflow://send?to=<beneficiaryId>&amount=<amount>&currency=<currency>
 *   remitflow://pay?ref=<paymentRef>
 *   remitflow://kyc
 *   remitflow://wallet
 *   remitflow://transaction/<id>
 *   https://app.remitflow.com/send?...  (Universal Link)
 *
 * Install: pnpm add @react-navigation/native (already included)
 */

import { Linking } from "react-native";
import { NavigationContainerRef } from "@react-navigation/native";

export type DeepLinkRoute =
  | { screen: "SendMoney"; params: { beneficiaryId?: string; amount?: string; currency?: string } }
  | { screen: "PaymentConfirm"; params: { ref: string } }
  | { screen: "KYC"; params: {} }
  | { screen: "Wallet"; params: {} }
  | { screen: "TransactionDetail"; params: { id: string } }
  | { screen: "Dashboard"; params: {} };

/**
 * Parse a deep link URL into a navigation route.
 */
export function parseDeepLink(url: string): DeepLinkRoute | null {
  try {
    // Normalize universal links to custom scheme
    const normalized = url
      .replace("https://app.remitflow.com", "remitflow://")
      .replace("https://remitflow.app", "remitflow://");

    const parsed = new URL(normalized);
    const path = parsed.pathname.replace(/^\//, "");
    const params = Object.fromEntries(parsed.searchParams.entries());

    if (path === "send" || parsed.host === "send") {
      return {
        screen: "SendMoney",
        params: {
          beneficiaryId: params.to,
          amount: params.amount,
          currency: params.currency ?? "NGN",
        },
      };
    }

    if (path === "pay" || parsed.host === "pay") {
      if (!params.ref) return null;
      return { screen: "PaymentConfirm", params: { ref: params.ref } };
    }

    if (path === "kyc" || parsed.host === "kyc") {
      return { screen: "KYC", params: {} };
    }

    if (path === "wallet" || parsed.host === "wallet") {
      return { screen: "Wallet", params: {} };
    }

    const txMatch = path.match(/^transaction\/(.+)$/);
    if (txMatch) {
      return { screen: "TransactionDetail", params: { id: txMatch[1] } };
    }

    return { screen: "Dashboard", params: {} };
  } catch {
    return null;
  }
}

/**
 * Navigate to the parsed deep link route.
 */
export function navigateToDeepLink(
  navigationRef: NavigationContainerRef<any>,
  route: DeepLinkRoute
): void {
  if (!navigationRef.isReady()) return;
  navigationRef.navigate(route.screen as never, route.params as never);
}

/**
 * Set up deep link listeners for foreground (app open) and cold start.
 * Call this in your root App component.
 */
export function setupDeepLinks(
  navigationRef: NavigationContainerRef<any>
): () => void {
  // Handle deep links when app is already open
  const subscription = Linking.addEventListener("url", ({ url }) => {
    const route = parseDeepLink(url);
    if (route) navigateToDeepLink(navigationRef, route);
  });

  // Handle cold start deep link
  Linking.getInitialURL().then((url) => {
    if (url) {
      const route = parseDeepLink(url);
      if (route) {
        // Delay navigation until navigator is ready
        const interval = setInterval(() => {
          if (navigationRef.isReady()) {
            navigateToDeepLink(navigationRef, route);
            clearInterval(interval);
          }
        }, 100);
        setTimeout(() => clearInterval(interval), 5000);
      }
    }
  });

  return () => subscription.remove();
}

/**
 * Generate a shareable deep link for sending money to a beneficiary.
 */
export function generateSendMoneyLink(
  beneficiaryId: string,
  amount?: number,
  currency = "NGN"
): string {
  const params = new URLSearchParams({ to: beneficiaryId });
  if (amount) params.set("amount", amount.toString());
  params.set("currency", currency);
  return `https://app.remitflow.com/send?${params.toString()}`;
}

/**
 * Generate a shareable payment request link.
 */
export function generatePaymentRequestLink(ref: string): string {
  return `https://app.remitflow.com/pay?ref=${encodeURIComponent(ref)}`;
}
